tmt/steps/prepare/__init__.py

```diff
@@ -22,40 +27,176 @@ def wake(self):
+            self.alias(i, 'script', 'playbook')
+            self.alias(i, 'script', 'path')
+            self.alias(i, 'script', 'inline')
+
+            self.debug('how', self)
+
+            for key, val in self.data[i].items():
+                if not val is None:
+                    self.info(f'{key}', val)
 
     def show(self):
         """ Show discover details """
         self.super.show(keys = ['how', 'script'])
 
     def go(self):
         """ Prepare the test step """
+        # TODO: not sure why
         self.super.go()
 
+        packages = []
+        for step in self.plan.steps():
+            if getattr(step, 'requires', False):
+                packages += step.requires()
+
+        if packages:
+            self.install(packages)
+
         for data in self.data:
-            how = data['how']
-            script = data['script']
+            self.run_how(data['how'], data['script'])
 
-            if script:
-                self.verbose('    Prepare', f"{how} = '{script}", 'yellow')
 
-                try:
-                    self.plan.provision.prepare(how, script)
-                except AttributeError as error:
-                    raise SpecificationError('NYI: cannot currently run this preparator.')
+    ## Knowhow ##
+    def run_how(self, how, what):
+        """ Run specific HOW """
+        self.info('Prepare', f"{how} = '{what}", 'yellow')
+        getattr(self,
+            f"how_{how}",
+            lambda: 'generic',
+            )(how, what)
 
-            else:
-                self.debug('Note', f"No path/script defined for prepare({how})", 'yellow')
+    def how_generic(self, how, what):
+        if not what:
+            self.debug('Note', f"No data provided for prepare({how})", 'yellow')
+            return
 
-        # TODO: find a better way
-        packages = self.plan.execute.requires()
-        if packages:
-            self.plan.provision.prepare('shell', f"sleep 1; set -x; nohup bash -c 'dnf install -y {' '.join(packages)}' 1>/root/prepare.log 2>&1 && exit 0; cat prepare.log; exit 1")
-            self.plan.provision.copy_from_guest('/root/prepare.log')
+        # Process multiple inputs
+        if type(what) is list:
+            return [self.how_generic(how, w) for w in what]
+
+        # Try guesssing the path
+        whatpath = os.path.join(self.plan.run.tree.root, what)
+
+        self.debug('Looking for prepare script in', whatpath)
+        if os.path.exists(whatpath) and os.path.isfile(whatpath):
+            what = whatpath
+        else:
+            whatpath = os.path.join(self.plan.workdir,
+                'discover',
+                self.data['name'],
+                'tests',
+                what)
+
+            self.debug('Looking for prepare script', whatpath)
+            if os.path.exists(whatpath) and os.path.isfile(whatpath):
+                what = whatpath
+
+        try:
+            self.plan.provision.prepare(how, what)
```

Comment:
Can we somehow improve prepare so it does not detect myriad of Linux distributions? I encountered this while looking why prepare phase takes too long:

```
[mvadkert@freedom /var/tmp/tmt/run-051/plans/basic/provision/gchBHnRSKzzDtZxm]$ time vagrant --debug provision --provision-with prepare
--- snip ---
DEBUG ssh: Checking whether SSH is ready...
DEBUG ssh: Re-using SSH connection.
 INFO ssh: SSH is ready!
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute:  (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 0
 INFO guest: Autodetecting host type for [#<Vagrant::Machine: default (VagrantPlugins::ManagedServers::Provider)>]
DEBUG guest: Trying: mint
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: if test -r /etc/os-release; then
source /etc/os-release && test 'xLinux Mint' = "x$ID" && exit
fi
if test -x /usr/bin/lsb_release; then
/usr/bin/lsb_release -i 2>/dev/null | grep -qi 'Linux Mint' && exit
fi
if test -r /etc/issue; then
cat /etc/issue | grep -qi 'Linux Mint' && exit
fi
exit 1
 (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: atomic
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: grep 'ostree=' /proc/cmdline (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: trisquel
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: [ -x /usr/bin/lsb_release ] && /usr/bin/lsb_release -i 2>/dev/null | grep Trisquel (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: elementary
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: if test -r /etc/os-release; then
source /etc/os-release && test 'xelementary' = "x$ID" && exit
fi
if test -x /usr/bin/lsb_release; then
/usr/bin/lsb_release -i 2>/dev/null | grep -qi 'elementary' && exit
fi
if test -r /etc/issue; then
cat /etc/issue | grep -qi 'elementary' && exit
fi
exit 1
 (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: pld
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: cat /etc/pld-release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: stderr: cat: /etc/pld-release: No such file or directory

DEBUG ssh: Exit status: 1
DEBUG guest: Trying: alt
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: cat /etc/altlinux-release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: stderr: cat: /etc/altlinux-release: No such file or directory

DEBUG ssh: Exit status: 1
DEBUG guest: Trying: funtoo
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: grep Funtoo /etc/gentoo-release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: stderr: grep: /etc/gentoo-release: No such file or directory

DEBUG ssh: Exit status: 2
DEBUG guest: Trying: kali
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: if test -r /etc/os-release; then
source /etc/os-release && test 'xkali' = "x$ID" && exit
fi
if test -x /usr/bin/lsb_release; then
/usr/bin/lsb_release -i 2>/dev/null | grep -qi 'kali' && exit
fi
if test -r /etc/issue; then
cat /etc/issue | grep -qi 'kali' && exit
fi
exit 1
 (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: dragonflybsd
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: uname -s | grep -i 'DragonFly' (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: ubuntu
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: if test -r /etc/os-release; then
source /etc/os-release && test 'xubuntu' = "x$ID" && exit
fi
if test -x /usr/bin/lsb_release; then
/usr/bin/lsb_release -i 2>/dev/null | grep -qi 'ubuntu' && exit
fi
if test -r /etc/issue; then
cat /etc/issue | grep -qi 'ubuntu' && exit
fi
exit 1
 (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: amazon
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: grep 'Amazon Linux' /etc/os-release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: fedora
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: grep 'Fedora release' /etc/redhat-release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: freebsd
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: uname -s | grep 'FreeBSD' (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: openbsd
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: uname -s | grep 'OpenBSD' (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: gentoo
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: grep Gentoo /etc/gentoo-release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: stderr: grep: /etc/gentoo-release: No such file or directory

DEBUG ssh: Exit status: 2
DEBUG guest: Trying: suse
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: test -f /etc/SuSE-release || grep -q SUSE /etc/os-release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: netbsd
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: uname -s | grep NetBSD (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: tinycore
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: if test -r /etc/os-release; then
source /etc/os-release && test 'xCore Linux' = "x$ID" && exit
fi
if test -x /usr/bin/lsb_release; then
/usr/bin/lsb_release -i 2>/dev/null | grep -qi 'Core Linux' && exit
fi
if test -r /etc/issue; then
cat /etc/issue | grep -qi 'Core Linux' && exit
fi
exit 1
 (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: photon
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: cat /etc/photon-release | grep 'VMware Photon' (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: stderr: cat: /etc/photon-release: No such file or directory

DEBUG ssh: Exit status: 1
DEBUG guest: Trying: solaris11
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: grep 'Solaris 11' /etc/release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: stderr: grep: /etc/release: No such file or directory

DEBUG ssh: Exit status: 2
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: uname -sr | grep 'SunOS 5.11' (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 1
DEBUG guest: Trying: redhat
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: cat /etc/redhat-release (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 0
 INFO guest: Detected: redhat!
... snip ...
```

Reply:
That's actually standard way Vagrant does things (on guest), not sure we even wan't to change those, but I'll look into it.

So it was a `time` issue? That is, it took a long time to execute it? Were you using remote guest?

Reply:
this simple prepare phase took for me from 1-3 minutes, while most of the time it was autodetecting stuff:

```
[mvadkert@freedom /var/tmp/tmt/run-051/plans/basic/provision/gchBHnRSKzzDtZxm]$ time vagrant --debug provision --provision-with prepare
real	1m47.346s
```

It would be awesome if we could say to vargrant this is a RHEL or Fedora machine, so he does not need to do the autodetection by himself.